pkg/vault/contracts/CompositeLiquidityRouter.sol

```diff
@@ -482,23 +486,76 @@ contract CompositeLiquidityRouter is ICompositeLiquidityRouter, BatchRouterCommo
+            }
+
             _currentSwapTokenInAmounts().tSet(tokensIn[i], params.maxAmountsIn[i]);
+            _currentSwapTokensIn().add(tokensIn[i]);
         }
 
-        IERC20[] memory parentPoolTokens = _vault.getPoolTokens(params.pool);
+        (uint256[] memory amountsIn, ) = _addLiquidity(params.pool, params);
+
+        // Adds liquidity to the parent pool, mints parentPool's BPT to the sender and checks the minimum BPT out.
+        (, exactBptAmountOut, ) = _vault.addLiquidity(
+            AddLiquidityParams({
+                pool: params.pool,
+                to: isStaticCall ? address(this) : params.sender,
+                maxAmountsIn: amountsIn,
+                minBptAmountOut: params.minBptAmountOut,
+                kind: params.kind,
+                userData: params.userData
+            })
+        );
+
+        // Settle the amounts in.
+        if (isStaticCall == false) {
+            _settlePaths(params.sender, params.wethIsEth);
+        }
+    }
+
+    function _addLiquidity(
+        address pool,
+        AddLiquidityHookParams calldata params
+    ) internal returns (uint256[] memory amountsIn, bool allAmountsEmpty) {
+        return _addLiquidityRecursive(pool, params, 1);
+    }
+
+    function _addLiquidityRecursive(
+        address currentPool,
+        AddLiquidityHookParams calldata params,
+        uint256 level
+    ) internal returns (uint256[] memory amountsIn, bool allAmountsEmpty) {
+        allAmountsEmpty = true;
+
+        IERC20[] memory parentPoolTokens = _vault.getPoolTokens(currentPool);
+        amountsIn = new uint256[](parentPoolTokens.length);
 
         // Iterate over each token of the parent pool. If it's a BPT, add liquidity unbalanced to it.
         for (uint256 i = 0; i < parentPoolTokens.length; i++) {
             address childToken = address(parentPoolTokens[i]);
 
+            // NOTE: Recursion involving a pool containing its own BPT is explicitly prohibited by the Vault.
+            // This case is impossible and does not require handling.
+
             if (_vault.isPoolRegistered(childToken)) {
                 // Token is a BPT, so add liquidity to the child pool.
 
-                IERC20[] memory childPoolTokens = _vault.getPoolTokens(childToken);
-                (uint256[] memory childPoolAmountsIn, bool childPoolAmountsEmpty) = _getPoolAmountsIn(
-                    childPoolTokens,
-                    params.sender,
-                    params.wethIsEth
+                if (level > _MAX_LEVEL_IN_NESTED_OPERATIONS) {
+                    // If we have exceeded the maximum level of nested operations,
+                    // the token will be considered a standard ERC20.
+                    if (_settledTokenAmounts().tGet(childToken) == 0) {
+                        amountsIn[i] = _currentSwapTokenInAmounts().tGet(childToken);
+                        _settledTokenAmounts().tSet(childToken, amountsIn[i]);
+                    }
+                    continue;
```

Comment:
Namely if settled token amounts is not 0 here it means that the token already appeared somewhere else. Shouldn't we revert there?

Also I wonder if there's a better way to do that, as settled token amounts means something different in the batch router.